### PR TITLE
WIP - create sync endpoint to flush blocks from memory to store

### PR DESCRIPTION
**Pull request by domgreen**
Usecase:
- would like to be able to scale up and down scrapers dynamically 
- do not want to lose data so want to flush / upload blocks that are in prom memory
- once this upload has been done we can allow the pod to be scaled down

### CODE REVIEW

**Comment by domgreen** [comment]
Work In Progress - pushing as I'm off for a few days.

My main concern is that we're adding a lot of cruft to `sidecar.go` and would prefer it not to be there but it does not IMO belong with the other http apis `query/api/v1.go` as the query would not do the snapshot and that I wouldnt want to pull in all the extra stuff in `v1`.

File: cmd/thanos/sidecar.go [hunk 1]
```diff
@@ -214,6 +229,70 @@ func runSidecar(
 		mux := http.NewServeMux()
 		registerMetrics(mux, reg)
 		registerProfile(mux)
+		if uploads {
+			mux.HandleFunc("/snapshot", func(w http.ResponseWriter, r *http.Request) {
```
Comment (Bplotka) [inline cmd/thanos/sidecar.go]: maybe it is time to actually move this to separate func (:

**Comment by mattbostock** [comment]
What's the use case for this please?

**Comment by domgreen** [comment]
@mattbostock updated description with the use case... TLDR - scale down scrapers without losing data.

**Comment by domgreen** [comment]
Discussion offline with @Bplotka main issues would be around uploading duplicate chunks ... We would want to ensure the shipper does not upload overlapping time periods when doing this.
We also could end up with a race condition where normal sync and finalising the scraper happen at similar times.so need to ensure this doesn't happen.

Will most likely move some logic into shipper and have a shared Thanos file for all uploads.

File: cmd/thanos/sidecar.go (hunk 1 above)
Comment (domgreen) [reply cmd/thanos/sidecar.go]: Agreed - will probably want to pass in a shared shipper as they will need to share some logic as well.

**Comment by antonio** [comment]
@domgreen is this being implemented in some other way? The use case looked exactly as the one problem I haven't figured out in my setup yet (recreating the sidecar pods because of a reschedule or deploy of the app usually lead to 2h gaps in the metrics data).

**Comment by domgreen** [comment]
@antonio would be good to some more info on your setup and use case 👍 

Closing it for the moment whilst we look deeper into how flushing the WAL may effect both Prometheus and Thanos for our scaling down of scrapers we will first go for a simplified approach of waiting till the next upload before scaling down the number of scrapers.

**Pull request closed without merging**